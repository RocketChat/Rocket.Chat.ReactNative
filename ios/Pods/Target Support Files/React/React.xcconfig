CLANG_CXX_LANGUAGE_STANDARD = c++14
CONFIGURATION_BUILD_DIR = ${PODS_CONFIGURATION_BUILD_DIR}/React
GCC_PREPROCESSOR_DEFINITIONS = $(inherited) COCOAPODS=1
HEADER_SEARCH_PATHS = $(inherited) "${PODS_ROOT}/Headers/Private" "${PODS_ROOT}/Headers/Private/React" "${PODS_ROOT}/Headers/Public" "${PODS_ROOT}/Headers/Public/React" "${PODS_ROOT}/Headers/Public/yoga" "$(PODS_TARGET_SRCROOT)/ReactCommon"
PODS_BUILD_DIR = ${BUILD_DIR}
PODS_CONFIGURATION_BUILD_DIR = ${PODS_BUILD_DIR}/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)
PODS_ROOT = ${SRCROOT}
PODS_TARGET_SRCROOT = ${PODS_ROOT}/../../node_modules/react-native
PRODUCT_BUNDLE_IDENTIFIER = org.cocoapods.${PRODUCT_NAME:rfc1034identifier}
SKIP_INSTALL = YES
