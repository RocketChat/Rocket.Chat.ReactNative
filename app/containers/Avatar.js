import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import FastImage from 'react-native-fast-image';
import Touch from '../utils/touch';

const formatUrl = (url, baseUrl, uriSize, avatarAuthURLFragment) => (
	`${ baseUrl }${ url }?format=png&width=${ uriSize }&height=${ uriSize }${ avatarAuthURLFragment }`
);

const Avatar = React.memo(({
	text, size, baseUrl, borderRadius, style, avatar, type, children, userId, token, onPress, theme
}) => {
	const avatarStyle = {
		width: size,
		height: size,
		borderRadius
	};

	if (!text && !avatar) {
		return null;
	}

	const room = type === 'd' ? text : `@${ text }`;

	// Avoid requesting several sizes by having only two sizes on cache
	const uriSize = size === 100 ? 100 : 50;

	let avatarAuthURLFragment = '';
	if (userId && token) {
		avatarAuthURLFragment = `&rc_token=${ token }&rc_uid=${ userId }`;
	}


	let uri;
	if (avatar) {
		uri = avatar.includes('http') ? avatar : formatUrl(avatar, baseUrl, uriSize, avatarAuthURLFragment);
	} else {
		uri = formatUrl(`/avatar/${ room }`, baseUrl, uriSize, avatarAuthURLFragment);
	}


	let image = (
		<FastImage
			style={avatarStyle}
			source={{
				uri,
				priority: FastImage.priority.high
			}}
		/>
	);

	if (onPress) {
		image = (
			<Touch onPress={onPress} theme={theme}>
				{image}
			</Touch>
		);
	}

	return (
		<View style={[avatarStyle, style]}>
			{image}
			{children}
		</View>
	);
});

Avatar.propTypes = {
	baseUrl: PropTypes.string.isRequired,
	style: PropTypes.any,
	text: PropTypes.string,
	avatar: PropTypes.string,
	size: PropTypes.number,
	borderRadius: PropTypes.number,
	type: PropTypes.string,
	children: PropTypes.object,
	userId: PropTypes.string,
	token: PropTypes.string,
	theme: PropTypes.string,
	onPress: PropTypes.func
};

Avatar.defaultProps = {
	text: '',
	size: 25,
	type: 'd',
	borderRadius: 4
};

export default Avatar;
