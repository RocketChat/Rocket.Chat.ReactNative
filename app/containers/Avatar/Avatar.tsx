import React from 'react';
import { View } from 'react-native';
import FastImage from 'react-native-fast-image';
import Touchable from 'react-native-platform-touchable';
import { settings as RocketChatSettings } from '@rocket.chat/sdk';

import { getAvatarURL } from '../../lib/methods/helpers/getAvatarUrl';
import { SubscriptionType } from '../../definitions';
import Emoji from '../markdown/Emoji';
import { IAvatar } from './interfaces';

const Avatar = React.memo(
	({
		server,
		style,
		avatar,
		children,
		userId,
		token,
		onPress,
		emoji,
		getCustomEmoji,
		avatarETag,
		isStatic,
		rid,
		blockUnauthenticatedAccess,
		serverVersion,
		text,
		size = 25,
		borderRadius = 4,
		type = SubscriptionType.DIRECT,
		externalProviderUrl
	}: IAvatar) => {
		if ((!text && !avatar && !emoji && !rid) || !server) {
			return null;
		}

		const avatarStyle = {
			width: size,
			height: size,
			borderRadius
		};

		let image;
		if (emoji) {
			image = <Emoji getCustomEmoji={getCustomEmoji} isMessageContainsOnlyEmoji literal={emoji} style={avatarStyle} />;
		} else {
			let uri = avatar;
			if (!isStatic) {
				uri = getAvatarURL({
					type,
					text,
					size,
					userId,
					token,
					avatar,
					server,
					avatarETag,
					serverVersion,
					rid,
					blockUnauthenticatedAccess,
					externalProviderUrl
				});
			}

			image = (
				<FastImage
					style={avatarStyle}
					source={{
						uri,
						headers: RocketChatSettings.customHeaders,
						priority: FastImage.priority.high
					}}
				/>
			);
		}

		if (onPress) {
			image = <Touchable onPress={onPress}>{image}</Touchable>;
		}

		return (
			<View style={[avatarStyle, style]} testID='avatar'>
				{image}
				{children}
			</View>
		);
	}
);

export default Avatar;
