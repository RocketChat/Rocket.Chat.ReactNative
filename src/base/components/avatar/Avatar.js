/**
 * React Native Avatar Component
 * the Component which show avatar
 *
 * @zack
 */

import React, { Component } from "react";
import {
  View,
  ViewPropTypes,
  Image,
  Text,
  TouchableOpacity,
  Dimensions,
  findNodeHandle,
  UIManager
} from "react-native";
import PropTypes from "prop-types";
import EStyleSheet from "react-native-extended-stylesheet";

import Lightbox from "../lightbox/Lightbox";

export default class Avatar extends Component {
  renderDefaultAvatar() {
    return (
      <View
        style={[
          styles.avatarStyle,
          { backgroundColor: "#B0C4DE" },
          this.props.avatarStyle
        ]}
      />
    );
  }

  renderAvatar() {
    const { width, height } = Dimensions.get("window");
    return this.props.lightbox ? (
      <Lightbox
        activeProps={{
          style: [styles.avatarImageActiveStyle, { width, height }]
        }}
      >
        <Image
          style={[styles.avatarStyle, this.props.avatarStyle]}
          source={{ uri: this.props.avatar }}
        />
      </Lightbox>
    ) : (
      <Image
        style={[styles.avatarStyle, this.props.avatarStyle]}
        source={{ uri: this.props.avatar }}
      />
    );
  }

  renderAvatarName() {
    if (this.props.showName == true) {
      var showThisName = this.props.name
        ? this.props.name
        : this.props.defaultName
        ? this.props.defaultName
        : "";
      return (
        <Text
          style={[styles.textStyle, this.props.textStyle]}
          allowFontScaling={false}
          numberOfLines={this.props.numberOfNameLines}
        >
          {showThisName}
        </Text>
      );
    }
    return null;
  }

  _avatarPress = () => {
    if (this.props.onAvatarPress) {
      this.props.onAvatarPress(this.props);
    }
  };

  layout(ref) {
    const handle = findNodeHandle(ref);

    return new Promise(resolve => {
      UIManager.measure(handle, (x, y, width, height, pageX, pageY) => {
        resolve({
          x,
          y,
          width,
          height,
          pageX,
          pageY
        });
      });
    });
  }

  render() {
    return (
      <TouchableOpacity
        style={[styles.containerStyle, this.props.avatarContainerStyle]}
        disabled={this.props.onAvatarPress ? false : true}
        onPress={this._avatarPress}
        accessibilityTraits="image"
      >
        {this.props.avatar ? this.renderAvatar() : this.renderDefaultAvatar()}
        {this.renderAvatarName()}
      </TouchableOpacity>
    );
  }
}

const styles = EStyleSheet.create({
  containerStyle: {
    flexDirection: "column",
    alignItems: "center"
  },
  avatarStyle: {
    justifyContent: "center",
    alignItems: "center",
    width: 40,
    height: 40,
    borderRadius: 5
  },
  avatarImageActiveStyle: {
    resizeMode: "contain"
  },
  textStyle: {
    color: "black",
    fontSize: 12,
    backgroundColor: "transparent",
    fontWeight: "normal"
  }
});

Avatar.defaultProps = {
  avatarStyle: { backgroundColor: "#faebd7" },
  showName: false,
  numberOfNameLines: 1,
  defaultName: "nothing"
};

Avatar.propTypes = {
  avatar: PropTypes.string,
  name: PropTypes.string,
  defaultName: PropTypes.string,
  avatarContainerStyle: ViewPropTypes.style,
  avatarStyle: ViewPropTypes.style,
  textStyle: Text.propTypes.style,
  onAvatarPress: PropTypes.func,
  showName: PropTypes.bool,
  numberOfNameLines: PropTypes.number,
  lightbox: PropTypes.bool
};
