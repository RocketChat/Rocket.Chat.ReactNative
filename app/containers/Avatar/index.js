import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Q } from '@nozbe/watermelondb';

import database from '../../lib/database';
import { getUserSelector } from '../../selectors/login';
import Avatar from './Avatar';

class AvatarContainer extends React.Component {
	static propTypes = {
		rid: PropTypes.string,
		text: PropTypes.string,
		type: PropTypes.string,
		blockUnauthenticatedAccess: PropTypes.bool,
		serverVersion: PropTypes.string
	};

	static defaultProps = {
		text: '',
		type: 'd'
	};

	constructor(props) {
		super(props);
		this.mounted = false;
		this.state = { avatarETag: '' };
		this.init();
	}

	componentDidMount() {
		this.mounted = true;
	}

	componentDidUpdate(prevProps) {
		const { text, type } = this.props;
		if (prevProps.text !== text || prevProps.type !== type) {
			this.init();
		}
	}

	componentWillUnmount() {
		if (this.subscription?.unsubscribe) {
			this.subscription.unsubscribe();
		}
	}

	get isDirect() {
		const { type } = this.props;
		return type === 'd';
	}

	init = async() => {
		const db = database.active;
		const usersCollection = db.get('users');
		const subsCollection = db.get('subscriptions');

		let record;
		try {
			if (this.isDirect) {
				const { text } = this.props;
				const [user] = await usersCollection.query(Q.where('username', text)).fetch();
				record = user;
			} else {
				const { rid } = this.props;
				record = await subsCollection.find(rid);
			}
		} catch {
			// Record not found
		}

		if (record) {
			const observable = record.observe();
			this.subscription = observable.subscribe((r) => {
				const { avatarETag } = r;
				if (this.mounted) {
					this.setState({ avatarETag });
				} else {
					this.state.avatarETag = avatarETag;
				}
			});
		}
	}

	render() {
		const { avatarETag } = this.state;
		const { serverVersion } = this.props;
		return (
			<Avatar
				avatarETag={avatarETag}
				serverVersion={serverVersion}
				{...this.props}
			/>
		);
	}
}

const mapStateToProps = state => ({
	user: getUserSelector(state),
	server: state.share.server.server || state.server.server,
	serverVersion: state.share.server.version || state.server.version,
	blockUnauthenticatedAccess:
		state.share.settings?.Accounts_AvatarBlockUnauthenticatedAccess
		?? state.settings.Accounts_AvatarBlockUnauthenticatedAccess
		?? true
});
export default connect(mapStateToProps)(AvatarContainer);
