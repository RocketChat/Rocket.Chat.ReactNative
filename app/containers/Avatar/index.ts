import Avatar from './AvatarContainer';

export { default as AvatarWithEdit } from './AvatarWithEdit';

export default Avatar;
