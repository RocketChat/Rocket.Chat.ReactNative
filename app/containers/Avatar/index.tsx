import React from 'react';
import { connect } from 'react-redux';
import { Q } from '@nozbe/watermelondb';
import { Observable, Subscription } from 'rxjs';

import database from '../../lib/database';
import { getUserSelector } from '../../selectors/login';
import { IApplicationState, TSubscriptionModel, TUserModel } from '../../definitions';
import Avatar from './Avatar';
import { IAvatar } from './interfaces';

class AvatarContainer extends React.Component<IAvatar, any> {
	private subscription?: Subscription;

	static defaultProps = {
		text: '',
		type: 'd'
	};

	constructor(props: IAvatar) {
		super(props);
		this.state = { avatarETag: '' };
		this.init();
	}

	componentDidUpdate(prevProps: IAvatar) {
		const { text, type } = this.props;
		if (prevProps.text !== text || prevProps.type !== type) {
			this.init();
		}
	}

	shouldComponentUpdate(nextProps: IAvatar, nextState: { avatarETag: string }) {
		const { avatarETag } = this.state;
		const { text, type, size, externalProviderUrl } = this.props;
		if (nextProps.externalProviderUrl !== externalProviderUrl) {
			return true;
		}
		if (nextState.avatarETag !== avatarETag) {
			return true;
		}
		if (nextProps.text !== text) {
			return true;
		}
		if (nextProps.type !== type) {
			return true;
		}
		if (nextProps.size !== size) {
			return true;
		}

		return false;
	}

	componentWillUnmount() {
		if (this.subscription?.unsubscribe) {
			this.subscription.unsubscribe();
		}
	}

	get isDirect() {
		const { type } = this.props;
		return type === 'd';
	}

	init = async () => {
		const db = database.active;
		const usersCollection = db.get('users');
		const subsCollection = db.get('subscriptions');

		let record;
		try {
			if (this.isDirect) {
				const { text } = this.props;
				const [user] = await usersCollection.query(Q.where('username', text)).fetch();
				record = user;
			} else {
				const { rid } = this.props;
				if (rid) {
					record = await subsCollection.find(rid);
				}
			}
		} catch {
			// Record not found
		}

		if (record) {
			const observable = record.observe() as Observable<TSubscriptionModel | TUserModel>;
			this.subscription = observable.subscribe(r => {
				const { avatarETag } = r;
				this.setState({ avatarETag });
			});
		}
	};

	render() {
		const { avatarETag } = this.state;
		const { serverVersion } = this.props;
		return <Avatar {...this.props} avatarETag={avatarETag} serverVersion={serverVersion} />;
	}
}

const mapStateToProps = (state: IApplicationState) => ({
	user: getUserSelector(state),
	server: state.share.server.server || state.server.server,
	serverVersion: state.share.server.version || state.server.version,
	blockUnauthenticatedAccess:
		(state.share.settings?.Accounts_AvatarBlockUnauthenticatedAccess as boolean) ??
		state.settings.Accounts_AvatarBlockUnauthenticatedAccess ??
		true,
	externalProviderUrl: state.settings.Accounts_AvatarExternalProviderUrl as string
});
export default connect(mapStateToProps)(AvatarContainer);
