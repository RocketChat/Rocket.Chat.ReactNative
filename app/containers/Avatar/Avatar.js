import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import FastImage from '@rocket.chat/react-native-fast-image';
import Touchable from 'react-native-platform-touchable';
import { settings as RocketChatSettings } from '@rocket.chat/sdk';

import { avatarURL } from '../../utils/avatar';
import Emoji from '../markdown/Emoji';

const Avatar = React.memo(({
	text,
	size,
	server,
	borderRadius,
	style,
	avatar,
	type,
	children,
	user,
	onPress,
	emoji,
	theme,
	getCustomEmoji,
	avatarETag,
	isStatic,
	rid
}) => {
	if ((!text && !avatar && !emoji && !rid) || !server) {
		return null;
	}

	const avatarStyle = {
		width: size,
		height: size,
		borderRadius
	};

	let image;
	if (emoji) {
		image = (
			<Emoji
				theme={theme}
				baseUrl={server}
				getCustomEmoji={getCustomEmoji}
				isMessageContainsOnlyEmoji
				literal={emoji}
				style={avatarStyle}
			/>
		);
	} else {
		let uri = avatar;
		if (!isStatic) {
			uri = avatarURL({
				type,
				text,
				size,
				user,
				avatar,
				server,
				avatarETag,
				rid
			});
		}

		image = (
			<FastImage
				style={avatarStyle}
				source={{
					uri,
					headers: RocketChatSettings.customHeaders,
					priority: FastImage.priority.high
				}}
			/>
		);
	}

	if (onPress) {
		image = (
			<Touchable onPress={onPress}>
				{image}
			</Touchable>
		);
	}


	return (
		<View style={[avatarStyle, style]}>
			{image}
			{children}
		</View>
	);
});

Avatar.propTypes = {
	server: PropTypes.string,
	style: PropTypes.any,
	text: PropTypes.string,
	avatar: PropTypes.string,
	emoji: PropTypes.string,
	size: PropTypes.number,
	borderRadius: PropTypes.number,
	type: PropTypes.string,
	children: PropTypes.object,
	user: PropTypes.shape({
		id: PropTypes.string,
		token: PropTypes.string
	}),
	theme: PropTypes.string,
	onPress: PropTypes.func,
	getCustomEmoji: PropTypes.func,
	avatarETag: PropTypes.string,
	isStatic: PropTypes.bool,
	rid: PropTypes.string
};

Avatar.defaultProps = {
	text: '',
	size: 25,
	type: 'd',
	borderRadius: 4
};

export default Avatar;
