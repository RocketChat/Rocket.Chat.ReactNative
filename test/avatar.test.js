import React from "react";
import Avatar from "../src/base/components/avatar/Avatar";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const avatarProps = {
    avatar: "https://jestjs.io/img/jest.svg", //'https://img3.doubanio.com/img/fmadmin/large/31905.jpg',
    name: "jestjs",
    showName: true,
    showNameLength: 100,
    avatarStyle: {
      width: 60,
      height: 60
    }
  };

  const tree = renderer.create(<Avatar {...avatarProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

// it('renders as an anchor when no page is set', () => {
//     const tree = renderer.create(<Link>Facebook</Link>).toJSON();
//     expect(tree).toMatchSnapshot();
// });

// it('properly escapes quotes', () => {
//     const tree = renderer
//         .create(<Link>{"\"Facebook\" \\'is \\ 'awesome'"}</Link>)
//         .toJSON();
//     expect(tree).toMatchSnapshot();
// });

// it('changes the class when hovered', () => {
//     const component = renderer.create(
//         <Link page="http://www.facebook.com">Facebook</Link>
//     );
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();

//     // manually trigger the callback
//     tree.props.onMouseEnter();
//     // re-rendering
//     tree = component.toJSON();
//     expect(tree).toMatchSnapshot();

//     // manually trigger the callback
//     tree.props.onMouseLeave();
//     // re-rendering
//     tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
// });
