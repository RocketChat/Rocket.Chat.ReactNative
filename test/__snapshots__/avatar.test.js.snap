// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityTraits="image"
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://jestjs.io/img/jest.svg",
      }
    }
    style={
      Array [
        undefined,
        Object {
          "height": 60,
          "width": 60,
        },
      ]
    }
  />
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    jestjs
  </Text>
</View>
`;
